### DISCUSSION TITLE
Does AWS SDK for Rust support middleware for service clients?

### DISCUSSION THREAD

**Discussion by pcgeek86** [awslabs/aws-sdk-rust#1203] [General]
The AWS SDK for JavaScript v3 supports the concept of "middleware" for service clients and specific service API operations (commands, as they call them). This functionality is useful for things like:

- Logging API calls elsewhere
- Creating session-level history of API calls
- Evaluating API parameters
- Manipulating resource tags
- Sanitizing request / response parameters
- Aborting the request, based on criteria

Does the AWS SDK for Rust support a similar concept of middleware / hooks / plugins?

https://aws.amazon.com/blogs/developer/middleware-stack-modular-aws-sdk-js/
https://github.com/aws/aws-sdk-js-v3/blob/main/README.md#middleware-stack

**Comment by aajtodd** [comment]
Yes, though the API's are different. The Rust SDK has the concept of [interceptors](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/interceptors.html) which is based on a common capability/API for all new AWS SDKs (you'll find the same interface in Kotlin + Swift for instance).

**Comment by pcgeek86** [comment]
Great, thank you. I had a feeling it was available. Just didn't know what it was called.
